### PR TITLE
Latest changes to agent and Reeti code.

### PR DESCRIPTION
**Pull request by mshayganfar**
The face tracking is working properly for the agent.

### CODE REVIEW

File: plugins/Plugins.Startup/App.xaml.cs [hunk 1]
```diff
@@ -26,9 +26,20 @@ public App()
 
         protected override void OnStartup(StartupEventArgs e)
         {
-			if ( e.Args.Length > 0 )
-				AgentControl.agentType = (AgentControl.AgentType)
-					Enum.Parse(typeof(AgentControl.AgentType), e.Args[0], true);
+            switch (e.Args[0])
```
Comment (charlesrich) [inline plugins/Plugins.Startup/App.xaml.cs]: Why did you change this from the compact version previously?

File: core/Client/Client.UI/AgentControl.xaml.cs [hunk 2]
```diff
@@ -32,8 +32,10 @@ namespace Agent.UI
 	public partial class AgentControl : UserControl, IAgentControl
 	{
         public enum AgentType { Unity, Reeti, Mirror };
+        
+        public static AgentType agentType;
 
-		public static AgentType agentType = AgentType.Unity;
```
Comment (charlesrich) [inline core/Client/Client.UI/AgentControl.xaml.cs:36]: I think it is better if the agentType has _default_ value of Unity?

File: core/agent/java/edu/wpi/always/client/GazeRealizer.java [hunk 3]
```diff
@@ -30,19 +30,30 @@ public static Point translateAgentTurn (float hor, float ver) {
...
    /**
-    * Convert camera image coordinates to horizontal
-    * component of agent coordinates.
+    * Convert camera image coordinates to horizontal component of agent
+    * coordinates.
     * 
     * @return between (inclusive) -1 (agent's right) and +1 (agent's left)
     */
-    public static float translateToAgentTurnHor (Point p) {
-       // NB: Make sure to change translateAgentTurn if change this
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/GazeRealizer.java:39]: You changed this, but you do not appear to have consistently changed translateAgentTurn, as per warning comment?

File: core/agent/java/edu/wpi/always/client/reeti/ReetiCommandSocketConnection.java [hunk 4]
```diff
@@ -1,19 +1,24 @@
...
 import java.net.*;
 
-
 public class ReetiCommandSocketConnection {
 
    private final Socket socket;
+
    private final PrintWriter writer;
 
+   public final ReetiJsonConfiguration reetiJsonConfig;
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/reeti/ReetiCommandSocketConnection.java]: This variable is not longer needed, since host is now passed into ReetCommandSocketConnection.

File: core/agent/java/edu/wpi/always/client/reeti/ReetiJsonConfiguration.java [hunk 5]
```diff
@@ -1,9 +1,169 @@
...
+         this.topLip = Double.valueOf((String) jsonObject.get("topLip"));
+
+         this.neckRotat = Double.valueOf((String) jsonObject.get("neckRotat"));
+
+         this.neckTilt = Double.valueOf((String) jsonObject.get("neckTilt"));
+
+         this.neckPan = Double.valueOf((String) jsonObject.get("neckPan"));
+
+      } catch (ParseException e) {
+         e.printStackTrace();
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/reeti/ReetiJsonConfiguration.java]: Better:  throw new RuntimeException(e);

File: core/agent/java/edu/wpi/always/client/reeti/ReetiPIDController.java [hunk 6]
```diff
@@ -1,6 +1,6 @@
-package edu.wpi.always.client;
+package edu.wpi.always.client.reeti;
 
-public class ReetiPIDController {
+public class ReetiPIDController extends ReetiJsonConfiguration {
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/reeti/ReetiPIDController.java]: For consistency, rather than reading the json file in again (which this will do), please follow the same strategy used elsewhere to get single copy of configuration object that is in PicoContainer (ask me to explain the dependency injection---I think I can do it very simply, and it is neat!)  Sorry I missed this use of config when I did the example code elsewhere.  I have updated in it develop, where you can update first before doing a new pull request.   Btw, there is a small possibility my updated code may have a problem, as I could not test it with Reeti.  Let me know if this is the case.

File: core/agent/java/edu/wpi/always/client/reeti/ReetiPIDMessages.java [hunk 7]
```diff
@@ -1,12 +1,14 @@
-package edu.wpi.always.client;
+package edu.wpi.always.client.reeti;
 
 import edu.wpi.always.client.*;
 
-public class ReetiPIDMessages {
+public class ReetiPIDMessages extends ReetiJsonConfiguration {
```
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/reeti/ReetiPIDMessages.java]: See my comment on ReetiPIDController.  Same issue.

File: plugins/rummy-plugin/rummy/.project [hunk 8]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (charlesrich) [inline plugins/rummy-plugin/rummy/.project]: rummy is obsolete (replaced by srummy).   Please delete it from your Eclipse (but do not delete files).

File: plugins/scriptbuilder-plugin/.classpath [hunk 9]
```diff
@@ -1,7 +1,7 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <classpath>
 	<classpathentry kind="src" path="java"/>
-	<classpathentry exported="true" kind="lib" path="scripts/AAScripts.jar"/>
 	<classpathentry combineaccessrules="false" kind="src" path="/agent"/>
+	<classpathentry exported="true" kind="lib" path="C:/Users/mel/Documents/GitHub/always/plugins/scriptbuilder-plugin/scripts/AAScripts.jar"/>
```
Comment (charlesrich) [inline plugins/scriptbuilder-plugin/.classpath]: No way we want this classpath in the repository!

File: user/User.owl [hunk 10]
```diff
@@ -43,7 +43,7 @@ ClassAssertion(<#User> <#Diane%20Ferguson>)
 ObjectPropertyAssertion(<#PersonLivesIn> <#Diane%20Ferguson> <#02118>)
 DataPropertyAssertion(<#PersonAge> <#Diane%20Ferguson> "62"^^xsd:integer)
 DataPropertyAssertion(<#PersonName> <#Diane%20Ferguson> "Diane Ferguson"^^xsd:string)
-DataPropertyAssertion(<#UserCloseness> <#Diane%20Ferguson> "Companion"^^xsd:string)
+DataPropertyAssertion(<#UserCloseness> <#Diane%20Ferguson> "Stranger"^^xsd:string)
```
Comment (charlesrich) [inline user/User.owl]: You are reverting this to an older version.   The current version should have Companion.

File: core/agent/java/edu/wpi/always/client/GazeRealizer.java (hunk 3 above)
Comment (mshayganfar) [inline core/agent/java/edu/wpi/always/client/GazeRealizer.java:39]: Do we need to change translateAgentTurn in this case? How can we check the potential problem?

Comment (mshayganfar) [inline core/agent/java/edu/wpi/always/client/GazeRealizer.java:39]: PS. based on the type of tuning we have, it is kind of over complicated.

File: core/agent/java/edu/wpi/always/client/reeti/ReetiPIDController.java (hunk 6 above)
Comment (mshayganfar) [inline core/agent/java/edu/wpi/always/client/reeti/ReetiPIDController.java]: Can we talk about this on Hangout for better understanding on Friday?

File: plugins/scriptbuilder-plugin/.classpath (hunk 9 above)
Comment (mshayganfar) [inline plugins/scriptbuilder-plugin/.classpath]: I don't know why this appeared! I have done nothing with AAScripts.jar!

File: core/agent/java/edu/wpi/always/client/GazeRealizer.java (hunk 3 above)
Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/GazeRealizer.java:39]: On Fri, Jan 17, 2014 at 12:56 AM, mshayganfar notifications@github.comwrote:

> In core/agent/java/edu/wpi/always/client/GazeRealizer.java:
> 
> > ```
> >  *
> >  * @return between (inclusive) -1 (agent's right) and +1 (agent's left)
> >  */
> > ```
> > -    public static float translateToAgentTurnHor (Point p) {
> > -       // NB: Make sure to change translateAgentTurn if change this
> 
> Do we need to change translateAgentTurn in this case?

Yes, these two functions must be inverse; translateAgentTurn is called 3
places in GazeBehavior, which you could easily verify (yourself!) in
Eclipse.

> How can we check the potential problem?

What do you mean by "potential problem"?

> —
> Reply to this email directly or view it on GitHubhttps://github.com/always-on/always/pull/196/files#r8953006
> .

## 

Dr. Charles Rich, Professor, Computer Science Departmenthttp://www.cs.wpi.edu
Interactive Media and Game Development
Programhttp://www.wpi.edu/academics/imgd/
Robotics Engineering Program http://www.wpi.edu/academics/robotics/
Worcester Polytechnic Institute, Fuller Laboratories B25b
100 Institute Road, Worcester, MA 01609-2280

Email: rich@wpi.edu   Phone: 508-831-5945   Fax: 508-831-5776
Home: http://www.cs.wpi.edu/~rich

Comment (charlesrich) [inline core/agent/java/edu/wpi/always/client/GazeRealizer.java:39]: On Fri, Jan 17, 2014 at 12:56 AM, mshayganfar notifications@github.comwrote:

> In core/agent/java/edu/wpi/always/client/GazeRealizer.java:
> 
> > ```
> >  *
> >  * @return between (inclusive) -1 (agent's right) and +1 (agent's left)
> >  */
> > ```
> > -    public static float translateToAgentTurnHor (Point p) {
> > -       // NB: Make sure to change translateAgentTurn if change this
> 
> PS. based on the type of tuning we have, it is kind of over complicated.

Yes the conditionals make it a bit complicated, but I am sure you can do
it.  It is only simple arithmetic, after all.   Just sit down with a piece
of paper and work it out the inverse function.

MAKE SURE YOU DO SOME TEST CASES, e.g., with a little main program in
GazeRealizer something like:

```
    public static void main (String[] args) {
       test(0, 0); test(0,.5); test(.5,0); test(.5,.5); test(.5,1);
```

test(1,.5) test(1,1);
           test(0, -0.5); ... etc.

```
    }

    private static void test (float hor, float ver) {
         Point point = translateToAgentTurn(hor, ver);
         System.out.println(point+": "+(hor -
```

translateAgentTurnHor(point))+" "+(ver - translateAgentTurn(point)));
       }

It is common practice to put test cases in main methods like this.

-CR

> —
> Reply to this email directly or view it on GitHubhttps://github.com/always-on/always/pull/196/files#r8953026
> .

## 

Dr. Charles Rich, Professor, Computer Science Departmenthttp://www.cs.wpi.edu
Interactive Media and Game Development
Programhttp://www.wpi.edu/academics/imgd/
Robotics Engineering Program http://www.wpi.edu/academics/robotics/
Worcester Polytechnic Institute, Fuller Laboratories B25b
100 Institute Road, Worcester, MA 01609-2280

Email: rich@wpi.edu   Phone: 508-831-5945   Fax: 508-831-5776
Home: http://www.cs.wpi.edu/~rich

**Comment by mshayganfar** [comment]
I'll make another pull request soon.
